src/utils/columnUtils.ts

```diff
@@ -1,132 +1,6 @@
-  let totalFrozenColumnWidth = 0;
-  const { rawGroupBy } = metrics;
-
-  const columns = metrics.rawColumns.map(metricsColumn => {
-    let width = getSpecifiedWidth(metricsColumn, metrics.columnWidths, metrics.viewportWidth);
-
-    if (width === undefined) {
-      unassignedColumnsCount++;
-    } else {
-      width = clampColumnWidth(width, metricsColumn, metrics.minColumnWidth);
-      allocatedWidths += width;
-    }
-
-    const column: IntermediateColumn = { ...metricsColumn, width };
-
-    if (rawGroupBy?.includes(column.key)) {
-      column.frozen = true;
-      column.rowGroup = true;
-    }
-
-    if (column.frozen) {
-      lastFrozenColumnIndex++;
-    }
-
-    return column;
-  });
-
-  columns.sort(({ key: aKey, frozen: frozenA }, { key: bKey, frozen: frozenB }) => {
-    // Sort select column first:
-    if (aKey === SELECT_COLUMN_KEY) return -1;
-    if (bKey === SELECT_COLUMN_KEY) return 1;
-
-    // Sort grouped columns second, following the groupBy order:
-    if (rawGroupBy?.includes(aKey)) {
-      if (rawGroupBy.includes(bKey)) {
-        return rawGroupBy.indexOf(aKey) - rawGroupBy.indexOf(bKey);
-      }
-      return -1;
-    }
-    if (rawGroupBy?.includes(bKey)) return 1;
-
-    // Sort frozen columns third:
-    if (frozenA) {
-      if (frozenB) return 0;
-      return -1;
-    }
-    if (frozenB) return 1;
-
-    // Sort other columns last:
-    return 0;
-  });
-
-  const unallocatedWidth = metrics.viewportWidth - allocatedWidths;
-  const unallocatedColumnWidth = Math.max(
-    Math.floor(unallocatedWidth / unassignedColumnsCount),
-    metrics.minColumnWidth
-  );
-
-  // Filter rawGroupBy and ignore keys that do not match the columns prop
-  const groupBy: string[] = [];
-  const calculatedColumns: CalculatedColumn<R, SR>[] = columns.map((column, idx) => {
-    // Every column should have a valid width as this stage
-    const width = column.width ?? clampColumnWidth(unallocatedColumnWidth, column, metrics.minColumnWidth);
-    const newColumn = {
-      ...column,
-      idx,
-      width,
-      left,
-      sortable: column.sortable ?? metrics.defaultSortable,
-      resizable: column.resizable ?? metrics.defaultResizable,
-      formatter: column.formatter ?? metrics.defaultFormatter
-    };
-
-    if (newColumn.rowGroup) {
-      groupBy.push(column.key);
-      newColumn.groupFormatter = column.groupFormatter ?? ToggleGroupFormatter;
-    }
-
-    totalWidth += width;
-    left += width;
-    return newColumn;
-  });
-
-  if (lastFrozenColumnIndex !== -1) {
-    const lastFrozenColumn = calculatedColumns[lastFrozenColumnIndex];
-    lastFrozenColumn.isLastFrozenColumn = true;
-    totalFrozenColumnWidth = lastFrozenColumn.left + lastFrozenColumn.width;
-  }
-
-  return {
-    columns: calculatedColumns,
-    lastFrozenColumnIndex,
-    totalFrozenColumnWidth,
-    totalColumnWidth: totalWidth,
-    groupBy
-  };
-}
-
-function getSpecifiedWidth<R, SR>(
+export function getSpecifiedWidth<R, SR>(
```

Comment:
I think we can move `getSpecifiedWidth` and `clampColumnWidth` into `useViewportColumns.ts`, they're not meant to be used for anything, so no need to "share" them.